### PR TITLE
[Enhancement] support partition rollup for string column in mv

### PR DESCRIPTION
**Pull request by murphyatwork**
Fixes #issue

support such mv:
```sql
create materialized view mv 
partition by date_trunc('month', dt)
as 
select str2date(str, '%Y-%m-%d') as dt from tbl;
```

but not support this kind of partition: 
- `partition by date_trunc('month', str2date(str, '%Y%m%d'))`

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:

- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.2
  - [x] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:heart_eyes: **pass** : 52 / 57 (91.23%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/analyzer/MaterializedViewPartitionFunctionChecker.java|11|14|78.57%|[73, 86, 91]|
|:large_blue_circle:|com/starrocks/sql/analyzer/PartitionExprAnalyzer.java|24|26|92.31%|[60, 61]|
|:large_blue_circle:|com/starrocks/sql/common/SyncPartitionUtils.java|14|14|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/MaterializedViewAnalyzer.java|3|3|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:heart_eyes: **pass** : 52 / 57 (91.23%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/analyzer/MaterializedViewPartitionFunctionChecker.java|11|14|78.57%|[73, 86, 91]|
|:large_blue_circle:|com/starrocks/sql/analyzer/PartitionExprAnalyzer.java|24|26|92.31%|[60, 61]|
|:large_blue_circle:|com/starrocks/sql/common/SyncPartitionUtils.java|14|14|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/MaterializedViewAnalyzer.java|3|3|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:heart_eyes: **pass** : 105 / 122 (86.07%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/connector/jdbc/MysqlSchemaResolver.java|0|3|00.00%|[155, 179, 205]|
|:large_blue_circle:|com/starrocks/qe/StmtExecutor.java|1|2|50.00%|[1952]|
|:large_blue_circle:|com/starrocks/catalog/PartitionKey.java|6|10|60.00%|[167, 168, 169, 170]|
|:large_blue_circle:|com/starrocks/sql/analyzer/MaterializedViewPartitionFunctionChecker.java|16|21|76.19%|[84, 90, 91, 97, 102]|
|:large_blue_circle:|com/starrocks/sql/common/SyncPartitionUtils.java|57|61|93.44%|[245, 246, 604, 608]|
|:large_blue_circle:|com/starrocks/sql/analyzer/PartitionExprAnalyzer.java|20|20|100.00%|[]|
|:large_blue_circle:|com/starrocks/qe/DDLStmtExecutor.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/MaterializedViewAnalyzer.java|4|4|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:heart_eyes: **pass** : 105 / 122 (86.07%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/connector/jdbc/MysqlSchemaResolver.java|0|3|00.00%|[155, 179, 205]|
|:large_blue_circle:|com/starrocks/qe/StmtExecutor.java|1|2|50.00%|[1952]|
|:large_blue_circle:|com/starrocks/catalog/PartitionKey.java|6|10|60.00%|[167, 168, 169, 170]|
|:large_blue_circle:|com/starrocks/sql/analyzer/MaterializedViewPartitionFunctionChecker.java|16|21|76.19%|[84, 90, 91, 97, 102]|
|:large_blue_circle:|com/starrocks/sql/common/SyncPartitionUtils.java|57|61|93.44%|[245, 246, 604, 608]|
|:large_blue_circle:|com/starrocks/sql/analyzer/PartitionExprAnalyzer.java|20|20|100.00%|[]|
|:large_blue_circle:|com/starrocks/qe/DDLStmtExecutor.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/MaterializedViewAnalyzer.java|4|4|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:heart_eyes: **pass** : 105 / 122 (86.07%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/connector/jdbc/MysqlSchemaResolver.java|0|3|00.00%|[155, 179, 205]|
|:large_blue_circle:|com/starrocks/qe/StmtExecutor.java|1|2|50.00%|[1952]|
|:large_blue_circle:|com/starrocks/catalog/PartitionKey.java|6|10|60.00%|[167, 168, 169, 170]|
|:large_blue_circle:|com/starrocks/sql/analyzer/MaterializedViewPartitionFunctionChecker.java|16|21|76.19%|[84, 90, 91, 97, 102]|
|:large_blue_circle:|com/starrocks/sql/common/SyncPartitionUtils.java|57|61|93.44%|[245, 246, 604, 608]|
|:large_blue_circle:|com/starrocks/sql/analyzer/PartitionExprAnalyzer.java|20|20|100.00%|[]|
|:large_blue_circle:|com/starrocks/qe/DDLStmtExecutor.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/MaterializedViewAnalyzer.java|4|4|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:heart_eyes: **pass** : 105 / 122 (86.07%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/connector/jdbc/MysqlSchemaResolver.java|0|3|00.00%|[155, 179, 205]|
|:large_blue_circle:|com/starrocks/qe/StmtExecutor.java|1|2|50.00%|[1952]|
|:large_blue_circle:|com/starrocks/catalog/PartitionKey.java|6|10|60.00%|[167, 168, 169, 170]|
|:large_blue_circle:|com/starrocks/sql/analyzer/MaterializedViewPartitionFunctionChecker.java|16|21|76.19%|[84, 90, 91, 97, 102]|
|:large_blue_circle:|com/starrocks/sql/common/SyncPartitionUtils.java|57|61|93.44%|[245, 246, 604, 608]|
|:large_blue_circle:|com/starrocks/sql/analyzer/PartitionExprAnalyzer.java|20|20|100.00%|[]|
|:large_blue_circle:|com/starrocks/qe/DDLStmtExecutor.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/MaterializedViewAnalyzer.java|4|4|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/sql/common/SyncPartitionUtils.java [hunk 1]
```diff
@@ -548,10 +567,11 @@ public static Set<String> getPartitionNamesByRangeWithPartitionLimit(Materialize
 
         int lastPartitionNum;
         if (partitionTTLNumber > 0 && isAutoRefresh && autoRefreshPartitionsLimit > 0) {
-            lastPartitionNum = Math.min(partitionTTLNumber, autoRefreshPartitionsLimit);;
+            lastPartitionNum = Math.min(partitionTTLNumber, autoRefreshPartitionsLimit);
+            ;
```
Comment (LiShuMing) [inline fe/fe-core/src/main/java/com/starrocks/sql/common/SyncPartitionUtils.java:571]: remove this?

File: fe/fe-core/src/main/java/com/starrocks/catalog/PartitionKey.java [hunk 2]
```diff
@@ -525,49 +533,20 @@ public static PartitionKey read(DataInput in) throws IOException {
...
-            for (int i = 0; i < types.size(); i++) {
-                if (!types.get(i).equals(partitionKey.types.get(i))) {
-                    return false;
-                }
-            }
-        }
-
-        return true;
+        PartitionKey that = (PartitionKey) o;
+        assert Objects.equals(types, that.types);
```
[the remark this answers is not in the document]
Comment (mofeiatwork) [reply fe/fe-core/src/main/java/com/starrocks/catalog/PartitionKey.java:544]: it's used to report exceptions when comparing DateLiteral with StringLiteral

File: fe/fe-core/src/main/java/com/starrocks/sql/analyzer/PartitionExprAnalyzer.java [hunk 3]
```diff
@@ -27,18 +27,49 @@
...
             FunctionCallExpr functionCallExpr = (FunctionCallExpr) expr;
             Function builtinFunction = null;
             Type targetColType = partitionSlotRef.getType();
             String functionName = functionCallExpr.getFnName().getFunction();
             if (functionName.equalsIgnoreCase(FunctionSet.DATE_TRUNC)) {
-                Type[] dateTruncType = {Type.VARCHAR, targetColType};
-                builtinFunction = Expr.getBuiltinFunction(functionCallExpr.getFnName().getFunction(),
-                        dateTruncType, Function.CompareMode.IS_IDENTICAL);
+                analyzeDateTruncFunction(functionCallExpr, partitionSlotRef);
+                return;
```
[the remark this answers is not in the document]
Comment (mofeiatwork) [reply fe/fe-core/src/main/java/com/starrocks/sql/analyzer/PartitionExprAnalyzer.java:70]: fixed.

File: fe/fe-core/src/main/java/com/starrocks/sql/common/SyncPartitionUtils.java (hunk 1 above)
Comment (mofeiatwork) [reply fe/fe-core/src/main/java/com/starrocks/sql/common/SyncPartitionUtils.java]: fixed

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:heart_eyes: **pass** : 106 / 123 (86.18%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/connector/jdbc/MysqlSchemaResolver.java|0|3|00.00%|[155, 179, 205]|
|:large_blue_circle:|com/starrocks/qe/StmtExecutor.java|1|2|50.00%|[1934]|
|:large_blue_circle:|com/starrocks/catalog/PartitionKey.java|6|10|60.00%|[167, 168, 169, 170]|
|:large_blue_circle:|com/starrocks/sql/analyzer/MaterializedViewPartitionFunctionChecker.java|16|21|76.19%|[84, 90, 91, 97, 102]|
|:large_blue_circle:|com/starrocks/sql/common/SyncPartitionUtils.java|57|61|93.44%|[245, 246, 603, 607]|
|:large_blue_circle:|com/starrocks/sql/analyzer/PartitionExprAnalyzer.java|21|21|100.00%|[]|
|:large_blue_circle:|com/starrocks/qe/DDLStmtExecutor.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/analyzer/MaterializedViewAnalyzer.java|4|4|100.00%|[]|

**Review by LiShuMing — APPROVED** [review]

File: fe/fe-core/src/test/java/com/starrocks/analysis/CreateMaterializedViewTest.java [hunk 4]
```diff
@@ -876,25 +876,50 @@ public void testPartitionWithFunction() {
...
+                    "\"replication_num\" = \"1\"\n" +
+                    ") " +
+                    "as select str2date(d,'%Y%m%d') as p,  a, b, c, d from jdbc0.partitioned_db0.tbl1;";
+            starRocksAssert.withMaterializedView(sql);
+        }
+
+        // rollup
+        {
+            String sql = "create materialized view mv_date_trunc_str2date " +
+                    "partition by date_trunc('month', p) " +
```
Comment (ABingHuang) [inline fe/fe-core/src/test/java/com/starrocks/analysis/CreateMaterializedViewTest.java:915]:
```suggestion
                    "partition by date_trunc('month', str2date(d,'%Y%m%d')) " +
```
do we support this?

Comment (mofeiatwork) [reply fe/fe-core/src/test/java/com/starrocks/analysis/CreateMaterializedViewTest.java:915]: it's not supported.

**Review by kangkaisen — APPROVED** [review]

**Review by satanson — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.2

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.1

**Pull request merged**